src/turtle/env/model/TestNode.d

```diff
@@ -0,0 +1,273 @@
+      data required by different test cases simply by writing directly into the
+      fake node's storage.)
+    * Implement the abstract `clear()` method, to remove all data.
+    * Implement the other abstract methods of NodeBase (see
+      swarm.node.model.NeoNode).
+
+    Copyright:
+        Copyright (c) 2015-2018 dunnhumby Germany GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module turtle.env.model.TestNode;
+
+import ocean.transition;
+import ocean.text.convert.Formatter;
+import turtle.env.model.Registry;
+import ocean.core.Verify;
+import swarm.node.model.NeoNode;
+import swarm.node.connection.ConnectionHandler;
+
+/*******************************************************************************
+
+    Abstract fake node for integration with turtle's registry of env additions.
+
+    Also includes methods for starting and stopping the fake node.
+
+    Note: this class and derivatives are only used when running tests which need
+    to *access* a node (the turtle env addition provides a fake node that can
+    be inspected and modified by test cases). It is not relevant when running
+    tests *on* a node implementation itself.
+
+    Params:
+        ConnHandler = node connection handler type
+
+*******************************************************************************/
+
+public abstract class TestNode ( ConnHandler : ISwarmConnectionHandler )
+    : NodeBase!(ConnHandler), ITurtleEnv
+{
+    import swarm.neo.AddrPort;
+    import ocean.io.device.File;
+    import ocean.core.Buffer;
+    import ocean.net.server.SelectListener;
+    import ocean.task.Scheduler;
+    import ocean.task.Task;
+    import turtle.env.Shell;
+    import Integer = ocean.text.convert.Integer_tango;
+    import ocean.io.select.client.model.ISelectClient : IAdvancedSelectClient;
+    import ocean.net.server.connection.IConnectionHandlerInfo;
+    import ocean.io.select.protocol.generic.ErrnoIOException;
+    import ocean.util.log.Logger;
+
+    /// Enum defining the possibles states of the fake node service.
+    private enum State
+    {
+        Init,
+        Running,
+        Stopped
+    }
+
+    /// State of the fake node service.
+    private State state;
+
+    /// Used to prevent creating multiple fake nodes of the same type.
+    static bool already_created = false;
+
+    /// Flag indicating that unhandled exceptions from the node must be printed
+    /// in test suite trace.
+    private bool log_errors = true;
+
+    /// Logger shared by all turtle nodes of this type.
+    static private Logger log;
+
+    /***************************************************************************
+
+        Static constructor.
+
+    ***************************************************************************/
+
+    static this ( )
+    {
+        log = Log.lookup("testnode");
```

Comment:
Hmmm. Is there a way for the fake node implementation to override the `testnode` log name? It's useful when testing with multiple types of test nodes.

Reply:
Yeah, I'll change it to one logger per instance.